Introduce a custom filesystem library

chandlerc opened the pull request:
The standard filesystem API lacks significant functionality, ranging from correct and secure creation of directories and files within them by using `openat` and avoiding [TOCTOU] issues, to support for filesystem locking.

[TOCTOU]: https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use

The LLVM filesystem library has more functionality, but uses an API that is increasingly diverging from the standard, and also fails to defend against TOCTOU.

This library is designed to carefully model the Unix or POSIX filesystem concepts of `openat` to avoid TOCTOU. However, it also tries to limit itself to an API subset that LLVM's filesystem library has also implemneted and so we have a strong reason to expect to be possible to port to Windows reasonably.

This PR included several benchmarks that show that this implementation is also faster for the majority of operations than the C++ standard library. The only places where there is a consistent regression is in recursively creating directories, and this is directly connected to the approach of using `openat` as the basis. Even there, while the wall time regresses, the cycles and instructions are significantly improved.

There are a number of operations not yet included here, I've focused on a core set of opening, closing, creating, and removing, and then adding those that I saw the current toolchain code using actively. I'll plan to expand the operations as needed going forward.

A follow-up PR that I'll finish polishing and send next ports `//toolchain/install` to consistently use this library and `std::filesystem::path` to both exercise the library and showcase its use. I'll be working systematically across the toolchain to converge all the code, extending this library as needed.

For reference, benchmark results on my macOS laptop:
https://gist.github.com/chandlerc/29d1f4d465a835b8be5174a48dad2e8f

Benchmark results on a Asahi Linux M1 Mac Mini:
https://gist.github.com/chandlerc/c42d43dd6b9b91746ab314b2afa152f7

Benchmark results on a Linux server with weirdly slow FS operations:
https://gist.github.com/chandlerc/48301a7383eb3972d53351b7e35e0561

chandlerc committed:
Introduce a custom filesystem library

The standard filesystem API lacks significant functionality, ranging
from correct and secure creation of directories and files within them by
using `openat` and avoiding TOCTOU issues, to support for filesystem
locking.

The LLVM filesystem library has more functionality, but uses an API that
is increasingly diverging from the standard, and also fails to defend
against TOCTOU.

This library is designed to carefully model the Unix / Posix filesystem
concepts of `openat` to avoid TOCTOU. However, it also tries to limit
itself to an API subset that LLVM's filesystem library has also
implemneted and so we have a strong reason to expect to be possible to
port to Windows reasonably.

I've also included several benchmarks that show that this implementation
is also dramatically faster for the overwhelming majority of operations
than the C++ standard library. The only places where there is
a regression is in recursively creating directories, and this is
directly connected to the approach used in the C++ standard library not
being secure against TOCTOU.

There are a number of operations not yet included here, I've focused on
a core set of opening, closing, creating, and removing, and then adding
those that I saw the current toolchain code using actively. I'll plan to
expand the operations as needed going forward.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+//
+// Because these are by far the most common parameters and they have unambiguous
+// names, the enumerators are also available directly within the namespace.
+enum class CreationFlags {
```
```suggestion
enum class CreateFlags {
```
Grammatically, we have "OpenFlags" so "CreateFlags" would better match I think. Or "CreationFlags" and "OpeningFlags", but the shorter seems better.

I probably would also not use `Flags` for a name that isn't bitflags that can be combined together (the values inside are flags but those are implementation details). Maybe `CreateOptions`, but not critical.. up to you.

chandlerc replied:
Yeah, I like that distinction. I think adopted and adjusted accordingly.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+//
+// These permissions can also be combined with the `|` operator, so `R_OK |
+// W_OK` checks for both read and write access.
+enum class AccessCheck : int {
```
Ironically, this one is bitflags but isn't called as such, even though the comment says it. Maybe `AccessCheckFlags`?
```suggestion
enum class AccessCheckFlags : int {
```

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+// Flags controlling which permissions should be checked in an `Access` call.
+//
+// These permissions can also be combined with the `|` operator, so `R_OK |
+// W_OK` checks for both read and write access.
```
```suggestion
// These permissions can also be combined with the `|` operator, so `AccessCheckFlags::Read |
// AccessCheckFlags::Write` checks for both read and write access.
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+                                  static_cast<int>(rhs));
+}
+
+// The underlying type that should be used to model the mode of a type.
```
I am not sure what "mode of a type" means without going to look at uses of this type in APIs, could this comment explain a bit more?

chandlerc replied:
Doh, a typo. Also a terrible comment. Made it much better and moved permission docs here.

danakj replied:
Thanks, the docs here are much better now.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  Socket = S_IFSOCK,
+
+  // Mask for the Unix-variant types to allow easy extraction.
+  Mask = S_IFMT,
```
```suggestion
  PosixMask = S_IFMT,
```
Maybe?

chandlerc replied:
I'm a bit torn on what to do here. Curious what you or others think.

Should we work to *consistently* tag non-portable names? That would suggest `UnixFifo`, `UnixCharDevice`, and `UnixMask` here. And probably `unix_uid()` and `unix_inode()` or similar below.

On one hand, this is nicely explicit. But I'm not sure it adds much readability as code using these eventually ends up in an `#ifdef` that only is compiled on one platform, and within that code it'll be pretty redundant and contextually clear...

(To be clear, happy to do it if that's the consensus, and happy to use whatever prefix works best given discussion of platform naming.)

danakj replied:
I think that I am okay with _not_ prefixing all those things, but here it is specifically "a mask to limit yourself to unix-like things" so it felt like it's part of its identity in a different way.

I guess it helps to think about the future shape of things here. When there are two platforms, we are going to have two `Mask` entries in ifdefs, I guess? Maybe it makes sense that you can unconditionally use a `Mask` then to get whatever things are valid on the target platform.

If there is only going to be a `Mask` on Unix and not on other platforms, then I think I might prefer the prefix still, for the same reason that it requires a comment explaining that relationship.

chandlerc replied:
Eh, the more I think about this the more I like prefixing everything. There really isn't that much.

danakj replied:
SGTM

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  RegularFile = S_IFREG,
+  SymbolicLink = S_IFLNK,
+
+  // Non-portable Unix-variant specific types.
```
Do you mean POSIX specific here?

chandlerc replied:
I had tried to switch to "Unix-like" instead of "POSIX" to distinguish from technically-POSIX-compliant vs. platforms that intend to use Unix-like APIs for filesystem access.

I've added an explicit note about this terminology in the namespace comment and a footnote to record the detail and subtlety here. I've also tried to go through and fix everything I see using "POSIX" or "Unix-variant" unless there is a really specific reason.

Let me know if the explanation makes sense and you're comfortable with this wording. Whatever term we end up with, I'll try to update things to be consistent.

danakj replied:
Thanks for the explanation in the footnote, that SGTM

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  BlockDevice = S_IFBLK,
+  Socket = S_IFSOCK,
+
+  // Mask for the Unix-variant types to allow easy extraction.
```
Same Q about POSIX here and throughout

chandlerc replied:
See above, hopefully now consistent at least and can update to stay consistent if the style changes.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+
+// Creates a temporary directory and returns a removing directory handle to it.
+//
+// Each directory created should be unique and newly created by the call. It is
```
```suggestion
// Each directory created will be unique and newly created by the call. It is
```
I believe this can be phrased as a promise of the API, rather than a condition to be met (perhaps by the caller)

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  SymbolicLink = S_IFLNK,
+
+  // Non-portable Unix-variant specific types.
+  FIFO = S_IFIFO,
```
```suggestion
  Fifo = S_IFIFO,
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  //
+  // For each, read is an octal value of `1`, write `2`, and execute `4`.
+  //
+  // Windows gracefully degrades to the effective permissions modeled using
```
TIL the USA uses only one `l` in modelled :)

chandlerc replied:
TIL that non-US uses two `l`s... =D

(But generally, I'm a terrible speller so I'm just going by what the tools complain about most often...)

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+// main `File` types: `ReadFileRef`, `WriteFileRef`, and `ReadWriteFileRef`.
+//
+// Objects using this type have access to an open file handle to a specific file
+// and expose operation on that open file. These operations may fail directly
```
```suggestion
// and expose operations on that open file. These operations may fail directly
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  auto operator=(FileRefBase&&) -> FileRefBase& = default;
+  auto operator=(const FileRefBase&) -> FileRefBase& = default;
+
+  // Computers the file status.
```
```suggestion
  // Computes the file status.
```

Though maybe, "Retrieves" or "Reads"?

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // Analogous to the Unix-like `fstat` call.
+  auto Stat() -> ErrorOr<FileStatus, FdError>;
+
+  // Methods to seek the current file position, with various semantics for the
```
Can we document which one does which semantics? In particular Seek and SeekTo read ambiguously (the first is relative, maybe?)

chandlerc replied:
I read `SeekTo` as requiring a position rather than a delta, and hoped that would be a sufficient distinction. I take it not?

Would `SeekToPosition` be better?

danakj replied:
`SeekFromBeginning` would parallel with `SeekFromEnd` and be even more clear for me.

chandlerc replied:
Done!

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  //
+  // On success, this returns a new slice from the start to the end of the
+  // successfully read bytes. These will always be located in the passed-in
+  // buffer, but not all of it may be filled. A partial read does not mean that
```
```suggestion
  // buffer, but not all of the buffer may be filled. A partial read does not mean that
```
This reads like not all of the returned slice may be filled, but I don't think it's saying that

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // provide a single, non-templated implementation.
+  auto Close() && -> ErrorOr<Success, FdError>;
+
+  // Note: this is a private API that should not be made public, and should only
+  // be used by the implementation of subclass destructors. It should also only
+  // be called for subclasses with *ownership* of the file reference, and is
+  // provided here as a single non-template implementation.
```
Let's mention that this calls Close, so you don't have to call both?

chandlerc replied:
Good suggestion, done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  File() = default;
+
+  // File objects are move-only as they model ownership.
+  File(const File&) = delete;
```
We should delete the `operator=(const File&)` too?

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+
+  // File objects are move-only as they model ownership.
+  File(const File&) = delete;
+  File(File&& arg) noexcept : FileRef<A>(arg.fd_) { arg.fd_ = -1; }
```
```suggestion
  File(File&& arg) noexcept : FileRef<A>(std::exchange(arg.fd_, -1)) {}
```

chandlerc replied:
How important is using `std::exchange` for these situations to you?

I work to avoid it because it seems wasteful in terms of compile time and debug run time -- requires instantiating 2 class types, 2 functions, materializing temporaries for the RHS, a local variable for temporary storage, calling 3 functions, etc. =/

Leaving all the `std::exchange` suggestions until we finish this discussion thread.

danakj replied:
I do find that it greatly improves showing the intent of what is going on, and increases confidence in future refactors/changes not introducing bugs.

(Maybe we could write a more compile-efficient version if some of that is accidental complexity?)

chandlerc replied:
Yeah, I see the benefit I think. Mostly really sad at the cost.

I don't have any good ideas for doing this without a weirdly disproportionate cost for the abstraction provided....

Closest I think still doesn't qualify as "good", see `CARBON_XCHG` here: https://cpp.compiler-explorer.com/z/Ys7nfrfo4

That also maybe shows some of the concern -- 2x the generated code before inlining, and requires two levels of inlining + more optimization passes to get back to the same code.

But I think I'm wrong to stress here. we use `std::max` and `std::min` with the same downsides. Switching to `std::exchange` for now.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  auto operator=(File&& arg) noexcept -> File& {
+    this->Destroy();
+    this->fd_ = arg.fd_;
+    arg.fd_ = -1;
```
```suggestion
    this->fd_ = std::exchange(arg.fd_, -1);
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // allow checked destruction and release of the file resources.
+  //
+  // See the base class for full documentation of the `Close` signature and how
+  // to handle errors.
```
I didn't see this documentation, just a Note on the function. Maybe the docs could live here?

chandlerc replied:
Doh. And sure, can put them here. The awkward thing is not having the signature ready to hand, but I can just put that in the comment.

danakj commented:
I've started reviewing this as josh is away. Noting it here since it's a huge PR and taking some time.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+              AccessCheck check = AccessCheck::Exists)
+      -> ErrorOr<bool, PathError>;
+
+  // Returns the `FileStatus` for the open directory.
```
Maybe "Reads" rather than "Returns" as this is performing a filesystem access?

chandlerc replied:
Good suggestion.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // Returns the `FileStatus` for the open directory.
+  auto Stat() -> ErrorOr<FileStatus, FdError>;
+
+  // Returns the `FileStatus` for the provided path (without opening it).
```
Same, with "Reads"?

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // symlinks and provide the status of the underlying file or directory.
+  auto Stat(std::filesystem::path path) -> ErrorOr<FileStatus, PathError>;
+
+  // Returns the `FileStatus` for the provided path (without opening it).
```
"Reads" as well?

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // symlinks, and instead will return the status of the symlink itself.
+  auto Lstat(std::filesystem::path path) -> ErrorOr<FileStatus, PathError>;
+
+  // Returns the contents of the symlink at the provided path.
```
"Reads" as well?

Maybe worth clarifying what it means by "contents" here given the return type being `std::string`. If it's a symlink to a file, it's not returning the file's contents, but the path to that file, right?

A symlink can't contain something other than a path, right? So could this return a `std::filesystem::path`?

chandlerc replied:
Symlinks can contain any null-terminated string they want... It doesn't have to be anything remotely valid as a "path". =/

The representation of these as paths has always bugged me a bit because of this... but maybe it shouldn't and I should switch to a path. Dunno...

I've at least updated the comment to be much more clear about what is going on here.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // defaults to error unless opening an existing file.
+  //
+  // If creating a file, the file is created with `creation_mode` which defaults
+  // to a restrictive `0600`. Additional flags can be provided to `flags` to
```
Maybe `ModeType` should document the format for `0600`? It's currently explained only on `FileStatus::permissions()`

chandlerc replied:
Yeah, moved the docs to `ModeType`.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // defaults to error unless opening an existing file.
+  //
+  // If creating a file, the file is created with `creation_mode` which defaults
+  // to a restrictive `0600`. Additional flags can be provided to `flags` to
```
> Additional flags can be provided to `flags`

OpenFlags is not a bitmask so this reads incorrectly to me. Maybe if the type is no longer "Flags" it won't lead to referring to it like you can pass multiple as well.

chandlerc replied:
It should be a bit mask, and thats how you can in theory pass more. I've updated it to do that. Does that work?

There aren't any interesting combinations currently, but I expect them in the fullness of time...

danakj replied:
Yep, thanks

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+
+  // Opens the provided path as a write-only file. Otherwise, behaves as
+  // `OpenReadOnly`. Note that write-access of the open file and the permissions
+  // in the creation mode are orthogonal.
```
Is this trying to say that `creation_mode` must include write access in order for this to succeed?

chandlerc replied:
No, the opposite -- the returned access is completely independent of the permissions.

I've tried to give a more full description of this in the `OpenReadOnly` comment and removed the discussion here to just defer to that common explanation. That seem reasonable?

danakj replied:
Okay, thanks. Checking that I understood the comment correctly: You can open a file without write perms, but for writing, and write to it, but nothing else will be able to write to it after it's closed (without changing perms)?

chandlerc replied:
> Okay, thanks. Checking that I understood the comment correctly: You can open a file without write perms, but for writing, and write to it, but nothing else will be able to write to it after it's closed (without changing perms)?

Even weirder -- another open for write will fail even before closed.

Only the file descriptor that created the file with write permissions (prior to the permission bits being set) will be able to write. Only `dup`-ing that descriptor (or something similar like passing it via Unix domain socket) can make another write path AFAIK.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+                     OpenFlags flags = OpenFlags::None)
+      -> ErrorOr<WriteFile, PathError>;
+
+  // Opens the provided path as a write-only file. Otherwise, behaves as
```
```suggestion
  // Opens the provided path as a read-and-write file. Otherwise, behaves as
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // is more suitable for directories. There are no extra flags that can be
+  // passed.
+  //
+  // As with other open routines, when creating a directory, only the leaf
+  // component can be created by the call to this routine.
```
Maybe OpenReadOnly should mention this too then?

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // process's control or produce an error due to an existing directory.
+  // However, no validation is done on any prefix path components leading to the
+  // leaf component created. When securely creating directories, the initial
+  // creation should typically have a single component from a known and opened
```
What does "known" entail here?

chandlerc replied:
I was trying to say, existing. I've done that now. =]

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // This is a convenience wrapper for opening the path, reading the returned
+  // file to a string, and closing it. Errors from any step are returned.
+  auto ReadFileToString(std::filesystem::path path)
+      -> ErrorOr<std::string, PathError>;
```
The error could also be an FdError, no? So should this be a more general Error? Or should FdError and PathError be combined?

chandlerc replied:
The error types are based on the set of context available in the error message, not which operation specifically encounters the error. Here, we'll have a path to attach for context even when the actual error occurs during reading from the file descriptor.

I tried to document this in the comments for `FdError` and `PathError`, but let me know if I can do a better job there (or they just hadn't come up yet).

danakj replied:
I guess this is okay as we don't really have a good way to branch on "was this an FdError or PathError or..." with how the error type system is set up (when you go to a more general error type, all you preserve is a string), so losing the type isn't changing anything in that regard.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // Errors from any step are returned.
+  auto WriteFileFromString(std::filesystem::path path, llvm::StringRef content,
+                           CreationFlags creation_flags = CreateAlways)
+      -> ErrorOr<Success, PathError>;
```
Same Q about FdError

chandlerc replied:
See above.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+      -> ErrorOr<Success, PathError>;
+
+  // Changes the current working directory to this directory.
+  auto Chdir() -> ErrorOr<Success, FdError>;
```
I wonder if this should be `ChdirToSelf`?

chandlerc replied:
Hmm... would `Fchdir` be too obscure? (that's what this actually is doing)

I was also trying to parallel `Stat()` and `Stat("path")`.

danakj replied:
I think `Fchdir` would be less obvious (I hadn't seen that name before). I guess the relationship with Stat makes sense, let's leave it.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+
+  // Creates a symlink at the provided path with the contents of `target`.
+  //
+  // Note that the target of a symlink is an arbitrary string and there is no
```
While it can be an arbitrary string, why not encode the intent by using `std::filesystem::path` here, without validating the path exists? Do we need the ability to write things that aren't representable in a `path`?

chandlerc replied:
See the above discussion that it seemed useful to me to emphasize that this we can't make assumptions about it...

For example, `std::filesystem::path` says it only models *syntactic* aspects of the path, but if there exist *syntactically* invalid paths, it seems like we should support even those here.

Or as another example, we'd need to clarify that it would be invalid to do any normalization of the path before writing it, even though `path` has direct support for normalization that it claims doesn't access the filesystem and should be preserve the full semantic meaning. For a symlink target, you can observe how mane `/`s are used, etc.

My views here likely stem from working on a synthetic filesystem where we routinely used symlinks as a way to communicate arbitrary string content more easily than a reading a file. Which is maybe a somewhat odd background / use case....

danakj replied:
> a synthetic filesystem where we routinely used symlinks as a way to communicate arbitrary string content more easily than a reading a file

I find this highly amusing.

Thinking about the common use case it's going to be to pull out the symlink string and immediately make a path, or to construct a path to store in the symlink. My instinct for modelling this in an API would be to make the default choices work with paths to meet the user where they will be, and provide an advanced "raw symlink" version that works with std::string only. But perhaps that is overkill here.

chandlerc replied:
I was hoping we wouldn't need both because `std::filesystem::path` is constructible from a `std::string`, so the caller doesn't even have to do something to put it into a path. Similarly for `/`.

When doing `Symlink`, you'd have to do `my_path.native()` instead of `my_path`. I'd hoped that was a low-enough hurdle to not need the split APIs.

danakj replied:
Maybe this is fine, I just get a bit sad thinking about users round tripping through types with string copies being involved in common cases.

chandlerc replied:
Closing this out, I don't think we'll have excess copies here... but if that comes up, we can revisit. Let's start out with the "raw" API for now.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // error checking on whether it exists or is sensible. Also, the target string
+  // set will be up to the first null byte in `target`, regardless of its
+  // `size`. This will not overwrite an existing symlink at the provided path.
+  auto Symlink(std::filesystem::path path, const std::string& target)
```
```suggestion
  auto Symlink(std::filesystem::path path, llvm::StringRef target)
```
If we don't use `path`, we should not force use of an owning `string`

danakj replied:
Ah, I see, you want a nul terminator in the body. WTB CStringRef.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // created directory with a potentially untrusted entity. Once some prefix has
+  // been redirected to an untrustworthy location, subsequent calls cannot
+  // detect and avoid this. When creating directories securely, start from a
+  // securely directly created `Dir` with limited access, and then create
```
```suggestion
  // directly created `Dir` with limited access, and then create
```
I don't know that "securely" is adding anything except confusion here. We're describing the secure process, so we don't need to ask the user to do extra secure unspecified things

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // The path must not be a directory. If the path is a symbolic link, the link
+  // will be removed, not the target. Models the behavior of `unlinkat(2)` on
+  // Unix-like platforms.
+  auto Unlink(std::filesystem::path path) -> ErrorOr<Success, PathError>;
```
Maybe consider `RmFile` instead of `Unlink`?

Or document that this is deleting/removing it from the filesystem.

chandlerc replied:
Added documentation.

Where I can, I'm trying to parallel the Unix-like nomenclature for familiarity.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // The provided path must name a directory. This removes all files and
+  // subdirectories contained within that named directory and then removes the
+  // directory itself once empty.
+  auto RmdirRecursively(std::filesystem::path path)
```
Maybe consider `RmdirAndSubtree` or `RmTree` [(like shutil)](https://docs.python.org/3/library/shutil.html)?

chandlerc replied:
Ooo, I like referencing shutil. But I feel like these aren't really separable words, so is `Rmtree` (paralleling `Rmdir`) OK?

danakj replied:
SGTM

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+class Dir : public DirRef {
+ public:
+  Dir() = default;
+  Dir(Dir&& arg) noexcept : DirRef(arg.dfd_) { arg.dfd_ = -1; }
```
```suggestion
  Dir(Dir&& arg) noexcept : DirRef(std::exchange(arg.dfd_, -1)) {}
```

danakj replied:
```suggestion

  // Dir objects are move-only as they model ownership.
  Dir(Dir&& arg) noexcept : DirRef(arg.dfd_) { arg.dfd_ = -1; }
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  auto operator=(Dir&& arg) noexcept -> Dir& {
+    Destroy();
+    dfd_ = arg.dfd_;
+    arg.dfd_ = -1;
```
```suggestion
    dfd_ = std::exchange(arg.dfd_, -1);
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+ public:
+  Dir() = default;
+  Dir(Dir&& arg) noexcept : DirRef(arg.dfd_) { arg.dfd_ = -1; }
+  Dir(const Dir&) = delete;
```
Also delete the copy operator=?

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  FileRef(FileRef&&) = default;
+  FileRef(const FileRef&) = default;
+  auto operator=(FileRef&&) -> FileRef& = default;
+  auto operator=(const FileRef&) -> FileRef& = default;
```
Noting that `DirRef` just omits writing any of these, which maybe we should also do here?

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  FileRefBase(FileRefBase&&) = default;
+  FileRefBase(const FileRefBase&) = default;
+  auto operator=(FileRefBase&&) -> FileRefBase& = default;
+  auto operator=(const FileRefBase&) -> FileRefBase& = default;
```
Noting that `DirRef` just omits writing any of these, which maybe we should also do here?

chandlerc replied:
Yeah, not sure why I had them.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  ~RemovingDir();
+  auto operator=(RemovingDir&& rhs) -> RemovingDir& = default;
+
+  auto abs_path() const -> const std::filesystem::path& { return abs_path_; }
```
```suggestion
  auto abs_path() const -> const std::filesystem::path& [[clang::lifetimebound]] { return abs_path_; }
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+ public:
+  RemovingDir() = default;
+  RemovingDir(RemovingDir&& arg) = default;
+  explicit RemovingDir(Dir d, std::filesystem::path abs_path)
```
This looks like an odd public API, you could just stick anything in `abs_path`. Is this meant to only be used from `MakeTmpDir`?

Otherwise, maybe the `abs_path` doesn't really fit as a member of this class and should be returned alongside it from `MakeTmpDir`?

chandlerc replied:
The only way to have a scoped removing directory is to create it with both the opened directory and an absolute path...

Or you need an opened directory + path + base-directory-reference with complex lifetimes.

I can make it private and only used by `MakeTmpDir`, but it seems like a generally useful utility.

I could have *just* an `abs_path` option and do the open internally, but generally the way you get into this doesn't want to use the `abs_path` to do the open (just like `MakeTmpDir` doesn't)`. I have added a comment about the somewhat unusual contract here, but open to suggestions on the best way to model this.

danakj replied:
I was hoping a bit more code would let us do the RmTree from the opened dir alone (pull the path from the dir handle), but I guess not

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+class RemovingDir : public Dir {
+ public:
+  RemovingDir() = default;
+  RemovingDir(RemovingDir&& arg) = default;
```
nit: maybe group the move ctor and operator= together?

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // An empty optional indicates it is unknown what kind of entry this is, while
+  // an engaged optional contains a bool indicating whether it is known a
+  // directory or known a non-directory entry.
+  auto is_dir() const -> std::optional<bool> {
```
Consider an enum with 3 values instead of `optional<bool>`?

chandlerc replied:
Ooof... even when an optional bool is the desired semantic? Yeah, probably cleaner.

danakj replied:
Yeah, even though it wasn't written exactly as such, `if (optbool && *optbool) gets rather awkward to work with.

(Carbon would presumably not have a similar problem, thanks to using pattern matching instead of bool conversions for optional values.)

chandlerc replied:
Sorry I failed to actually implement my "Yeah" here on the previous iteration.

However, when implementing this I realized this one is awkward as an enum -- we don't know the set of possible values really. Even the latest manual page for Linux fails to list one possible value on my system. The result is that the enum would need to have a fixed size to support non-listed values. But even that is problematic as there is no real spec for what the base size would be... It's a `uint8_t` on Linux, but that seems bad to rely on.

Instead, I think this is pushing that the "right" way to code against this API is with predicates, but we need to bake into them the known-ness. I've switched to an explicit unknown predicate and added the `known` word to the other predicates. I think this matches how callers need to test this value, and at least avoids the awkwardness of `if (optbool & *optbol)`. WDYT?

danakj replied:
SGTM

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+    return entry_.dent_ == rhs.entry_.dent_;
+  }
+
+  auto operator*() const -> const Entry& { return entry_; }
```
```suggestion
  auto operator*() const -> const Entry& [[clang::lifetimebound]] { return entry_; }
```

That said, `Entry` just contains a pointer, so maybe return a copy?

Also, `CARBON_DCHECK(entry_.dent_)` to catch dereffing `end()`?

chandlerc replied:
My understanding is that this is required to return a reference to conform with the iterator API requirements, which is why I have an embedded object.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  Reader(Reader&& arg) noexcept : DirRef(arg.dfd_), dirp_(arg.dirp_) {
+    arg.dirp_ = nullptr;
+  }
+  Reader(const Reader&) = delete;
```
Also delete the copy operator=? Or just omit both? (The "Rule of 0, 3, or 5")

chandlerc replied:
I should just remove this to take us back to rule of 3. The base class handles this.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  Reader() = default;
+  Reader(Reader&& arg) noexcept : DirRef(arg.dfd_), dirp_(arg.dirp_) {
+    arg.dirp_ = nullptr;
+  }
```
```suggestion
  Reader(Reader&& arg) noexcept : DirRef(std::exchange(arg.dfd_, -1)), dirp_(std::exchange(arg.dirp_, nullptr)) {}
```

`Dir` moves the `dfd_` so `Reader` should do the same, right?

chandlerc replied:
Technically, its just `dirp_` that has ownership here, the `dfd_` is essentially just a cache of `dirfd(dirp_)` to read the fd back out of thu underlying directory struct.

But no reason to leave `dfd_` in a non-`-1` state after moving out of the arg. That said, doing it without `std::exchange` pending discussion of that.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+    Destroy();
+    dfd_ = arg.dfd_;
+    dirp_ = arg.dirp_;
+    arg.dirp_ = nullptr;
```
```suggestion
    dirp_ = std::exchange(arg.dirp_, nullptr);
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  Reader(const Reader&) = delete;
+  auto operator=(Reader&& arg) noexcept -> Reader& {
+    Destroy();
+    dfd_ = arg.dfd_;
```
```suggestion
    dfd_ = std::exchange(arg.dfd_, -1);
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // Accessor for the `errno` based error number. This is not a portable API,
+  // code using it will need to be ported to use a different API on Windows.
+  // TODO: Add a Windows-specific API for its low-level error information.
+  auto number() const -> int { return number_; }
```
Consider putting `posix` in its name? `posix_errno`?

chandlerc replied:
Will follow the prefix or not based on discussion above.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+
+ protected:
+  // NOLINTNEXTLINE(bugprone-crtp-constructor-accessibility):
+  explicit ErrnoErrorBase(int errnum) : number_(errnum) {}
```
```suggestion
  explicit ErrnoErrorBase(int errno) : number_(errno) {}
```

danakj replied:
Nevermind, this would shadow the real errno :)

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // Put the file in a moved-from state immediately as it is invalid to
+  // retry closing or use the file in any way even if the close fails.
+  int fd = fd_;
+  fd_ = -1;
```
```suggestion
  int fd = std::exchange(fd_, -1);
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+    auto result = std::move(*this).Close();
+    // Debug-only error check. For a non-debug build, we can just leak the
+    // file descriptor.
+    CARBON_DCHECK(result.ok(), "{0}", result.error());
```
This can cause flakiness in debug builds, is that a good idea?

chandlerc replied:
I'm torn...

On one hand, I agree, and would rather this just a normal check. There certainly isn't a *performance* concern here.

But this also is very unlikely to corrupt memory or otherwise make it difficult for the program to continue executing. Crashing seems unfortunately severe in that case.

Not sure, open to ideas or suggestions.

danakj replied:
I was thinking that we should not have a CHECK at all. If the user wants to discard errors then that's what we should do. I don't think a flaky-crashy (since it could occur quite transiently in a CI system for example) filesystem abstraction is doing them any favours.

Let's take this convo to the larger thread below. But I think we should do the same thing here that we do on closing a directory, whatever it is.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+    // file descriptor.
+    CARBON_DCHECK(result.ok(), "{0}", result.error());
+  }
+  fd_ = -1;
```
```suggestion
```

Close() already did this, unless the fd was already -1?

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // without retaining the original `Dir`.
+  //
+  // For more details about `Read`, see the documentation on `DirRef::Read`.
+  auto Read() && -> ErrorOr<Reader, FdError>;
```
Maybe consider giving this a unique name instead of overloading? That would make uses of this, for which the rvalue overload is sorta loadbearing/intentionally making different behaviour, more clear.

chandlerc replied:
Good idea, renamed to `TakeAndRead`. Happy to switch to a better name if one occurs to you.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // contents.
+  constexpr ssize_t BufferSize = 256;
+  char buffer[BufferSize];
+  ssize_t result = readlinkat(dfd_, path.c_str(), buffer, BufferSize);
```
```suggestion
  ssize_t read_bytes = readlinkat(dfd_, path.c_str(), buffer, BufferSize);
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+                            const std::string& target)
+    -> ErrorOr<Success, PathError> {
+  int result = symlinkat(target.c_str(), dfd_, path.c_str());
+  if (result == -1) {
```
```suggestion
  if (symlinkat(target.c_str(), dfd_, path.c_str()) == -1) {
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+}
+
+template <OpenAccess A>
+inline auto DirRef::OpenImpl(std::filesystem::path path,
```
Question throughout: It seems that copying a `path` would involve heap allocations, how come we receive by value instead of by const-ref in fn parameters? It may save a copy on errors, but it can add copies in success paths if the caller doesn't give an rvalue?

chandlerc replied:
I've added a discussion of this to the top-level comment on the namespace to try and explain. Happy to dig into it and whether it makes sense.

danakj replied:
SGTM. I guess you tried taking a const ref (and copying in the error case) and found it was not a perf improvement?

chandlerc replied:
No, mostly I found no perf difference at all when I switched back and forth once.

But that reminds me that I'm missing an important benchmark. Lemme write that and measure this. Maybe I'm wrong and actually copying in the error path is better.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+
+constexpr Dir::~Dir() { Destroy(); }
+
+inline auto Dir::Read() && -> ErrorOr<Reader, FdError> {
```
Why do these non-templated method bodies live in the header instead of the cpp file?

chandlerc replied:
Roughly, I've tried to only outline the functions that involve multiple system calls or multiple loops or complex logic. Most of these seem likely to be very good inlining candidates.

Ideally, a lot of this gets compiled down to just the straight line code that would result from directly using the syscall APIs.

Particularly for the richer error APIs, with inlining this is likely to have no material overhead compared to just testing `errno`, but without inlining we would do quite a bit of work to encode errors even when they aren't printed and are instead directly handled by calling code. Given the sometimes _expected_ nature of filesystem errors like `ENOENT`, that seemed important to me.

danakj replied:
Thanks for the reasoning

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  // Transition our file descriptor into a directory stream, clearing it in the
+  // process.
+  int dfd = dfd_;
+  dfd_ = -1;
```
```suggestion
  int dfd = std::exchange(dfd_, -1);
```

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  return Dir::Reader(dirp);
+}
+
+inline Dir::Dir(RemovingDir&& arg) noexcept : Dir(static_cast<Dir&&>(arg)) {}
```
RemovingDir has a member which won't be emptied though the stuff in the base classes will. (Elsewhere I wondered if it should keep that abs_path member but...) should we drop its contents here since we're slicing it off and moving the rest?

chandlerc replied:
The explicit goal of this is to drop the `abs_path_` and the removing semantics. I've tried to document the oddity of this in the comment on its declaration. The only caller should be the `Release` method on `RemovingDir`.

danakj replied:
Right, my concern is that if a user engages in use-after-move on the `RemovingDir` after `Release` they will have a half-functional thing, where the Dir parts are moved away but the `abs_path_` is still there and visibly so. Better to clear out everything in the moved-from `RemovingDir` consistently? Alternatively, take the `arg` by value which would do the same thing, and is maybe more representative of the desired semantics (always move-from, rather than conditionally move-from)?

Maybe this is where you tell me a moved-from `path` with a string in its SSO isn't emptied after move or something sad like that. "valid but unspecified" is not my friend. Regardless it feels buggy/incorrect to "move" the `arg` from the caller but then only move some of its fields.

chandlerc replied:
Oh, OH! I think I misunderstood your comment -- you're saying that we should *also* clear out the path so the `RemovingDir` is back to a fully empty state rather than leaving half of it? That makes lots of sense.

(And not at all worried about the cost of that, so no sad today!)

danakj replied:
Yeah, great, thanks

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+inline RemovingDir::~RemovingDir() {
+  if (dfd_ != -1) {
+    auto result = std::move(*this).Remove();
+    CARBON_CHECK(result.ok(), "{0}", result.error());
```
What makes us believe the filesystem can't fail here?

chandlerc replied:
See above, I think these are tightly coupled questions.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+constexpr auto Dir::Destroy() -> void {
+  if (dfd_ != -1 && dfd_ != AT_FDCWD) {
+    auto result = close(dfd_);
+    // Closing a directory shouldn't produce errors, directly check fail on any.
```
Why can't the filesystem produce an error here if it had something to flush?

chandlerc replied:
Directories, AFAIK, don't have the deferred error behavior of files on Unix-like platforms. For example, `closedir` has to close a directory file descriptor in addition to deallocating the struct used for `readdir`. Yet, it is clearly documented as only allowed to fail with `EBADF` on Linux -- https://man7.org/linux/man-pages/man3/closedir.3.html -- and `EBADF` and `EINTR` even in POSIX -- https://pubs.opengroup.org/onlinepubs/9799919799/functions/closedir.html

`EBADF` is a programming error, and one this library should make that essentially impossible.

There is the possibility of `EINTR` that I'm torn about, but it isn't clear what, if anything, we should do in the face of it:
https://man7.org/linux/man-pages/man2/close.2.html#:~:text=Dealing%20with%20error%20returns%20from%20close()

(AFAIK, this applies the same to `closedir`'s ability to produce `EINTR`, but its even more problematic to retry anything...)

Linux at least doesn't produce EINTR here precisely because it can't be retried. I'm tempted to keep not retrying and check fail to figure out if there is a system where we actually have to deal with this.

danakj replied:
Hm, thanks for the links.

Just to add to the discussion, here's freebsd's which also includes `ENOSPC`, but also which can not be gracefully handled: https://man.freebsd.org/cgi/man.cgi?close(2)

If we're going to keep the CHECKs on the basis of looking for platforms that can fail, under the assumption that the platforms we support _cannot_ fail with an error, then that's cool and let's document that, to be clear that if it ever does CHECK-fail, it's our fault and we need to handle it properly for that platform.

But also it seems clear that there are reasonable platforms that can fail - and we should not crash but also can't do anything with it, so we should just discard the errors as the user has requested by doing Destroy without an explicit Sync/Close first (even if we don't provide those for a Dir right now).

The exception seems to be platforms that can produce an error and allow retry, such as HP-UX, but there's no good way to CHECK for those. So we can leave a comment about it if we'd like, to carefully check a given platform for retryability when porting?

chandlerc replied:
> Hm, thanks for the links.
> 
> Just to add to the discussion, here's freebsd's which also includes `ENOSPC`, but also which can not be gracefully handled: [man.freebsd.org/cgi/man.cgi?close(2)](https://man.freebsd.org/cgi/man.cgi?close(2))

Hmm, that's `close` though, not `closedir`.

But FreeBSD's `closedir` is ... surprising:
https://man.freebsd.org/cgi/man.cgi?query=closedir#:~:text=The%20closedir()%20function%20may%20also%09fail%20and%20set%20errno%20for%20any%20of%20the%20%20er%2D%0A%20%20%20%20%20%20%20rors%20specified%20for%20the%20routine%20close(2).

So it claims it can fail with errors from `close(2)`, but then goes on below that to claim that `closedir` confirms to POSIX. And POSIX doesn't allow `ENOSPC` here. What does that mean? No idea!!!

More fundamentally, I think there is a reason *why* closing a directory FD doesn't produce deferred errors -- the "writes" aren't done through the file descriptor at all. Even when using `openat` to create a new entry, the directory file descriptor is used for _path resolution_, not the actual creation. Creation failures are required to be immediately provided from all of these routines, and there are none of the "data may not yet be written" for things that add entries to directories. The FD isn't opened with W vs. R permissions, etc.

> If we're going to keep the CHECKs on the basis of looking for platforms that can fail, under the assumption that the platforms we support _cannot_ fail with an error, then that's cool and let's document that, to be clear that if it ever does CHECK-fail, it's our fault and we need to handle it properly for that platform.

FWIW, this is my stance for *directories* specifically.

I think the discussion for regular files is very different. There, what I want is to discourage the use of the destructor outside of tests entirely... I'm not saying my DCHECK is a great way of achieving that, but was my goal...

> The exception seems to be platforms that can produce an error and allow retry, such as HP-UX, but there's no good way to CHECK for those. So we can leave a comment about it if we'd like, to carefully check a given platform for retryability when porting?

If we ever want to support HP-UX that does the retry thing (I haven't gone looking for the docs on the retry case specifically), I would do so by having dedicated code exclusively for that OS that detects and does allowed retries. But I'm hoping we can cross that bridge if/when we need to support that platform.

danakj replied:
Ok so to summarize where we are at:

- For directories, we agree that we can CHECK that no error occurs. Let's document that on porting the code to new platforms, care should be taken to ensure no errors are possible, or to retry on error if the platform supports it.
- For files, I don't think we should DCHECK as if the intent is to use that path in tests, it's going to cause tests to be flaky.

If we want code to *not* call destructor without explicit Close, we can track a bool and CHECK that Close was called in the dtor. If want code to be *allowed to* call destructor without explicit Close, then we should do nothing in the dtor and drop errors.

chandlerc replied:
> Ok so to summarize where we are at:
> 
> * For directories, we agree that we can CHECK that no error occurs. Let's document that on porting the code to new platforms, care should be taken to ensure no errors are possible, or to retry on error if the platform supports it.

LG and hopefully done.

> * For files, I don't think we should DCHECK as if the intent is to use that path in tests, it's going to cause tests to be flaky.
> 
> If we want code to _not_ call destructor without explicit Close, we can track a bool and CHECK that Close was called in the dtor. If want code to be _allowed to_ call destructor without explicit Close, then we should do nothing in the dtor and drop errors.

Hmm. And this is tricky because I want *test* code to be able to just call the destructor, and non-test code to call `Close`. Which seems hard to satisfy. I don't think we want two types to model the two things...

Maybe the best thing is to require `Close` to be explicitly called and have a helper on `ErrorOr` like `.Check()` so you can write `f.Close().Check()`?

danakj replied:
That sounds pretty nice, yes.

zygoloid replied:
(I was asked to offer an opinion here. It sounds like the question on directories has been answered, so I'm just going to consider files.)

Unless you're writing to the file, requiring an explicit `Close` seems like an unnecessary burden to me, as there's no risk of data loss or other badness in that case. I think the worst case is that you're on an uncommon OS where a `close` failure leaves the FD open, in which case there's not really anything you can do anyway.

If you *are* writing to the file and the `close` syscall fails, it would definitely be nice to notify the caller of this so that they can somehow report to the user that the output file is potentially corrupted and terminate the program cleanly. And so it'd be nice to require that API users who are writing to the file explicitly call `Close` themselves and handle the error it produces.

I suppose we could split the `File` and `FileRef` interfaces into read-only and read-write interfaces, and only require an explicit `Close` for a writable file? But that would add a fair amount of extra complexity. The other alternative that makes sense to me would be to silently ignore `Close` failures here, and document that you need to call `Close` manually if you want to do something other than silently ignore close failures. (Could we add a `bool` to `File` in an `#ifndef NDEBUG` that tracks whether it might have pending writes, and check-fail if it was destroyed with the bool set to true?)

chandlerc replied:
> I suppose we could split the `File` and `FileRef` interfaces into read-only and read-write interfaces, and only require an explicit `Close` for a writable file?

We actually already have that split: only `WriteFile` and `ReadWriteFile` have the writing APIs enabled. Would be easy to only require a close on those, and let the `ReadFile` just drop `close` errors on destruction.

Maybe that, plus something on `ErrorOr` to easy checking (in the `ErrorOr<Success, ...>` case)?

zygoloid replied:
Ah, perfect! Sorry, I was just quickly looking at the `FileRefBase` API and missed that the other types have this fanciness :)

chandlerc replied:
Just closing out this thread -- implemented the suggested semantics. Writable files require explicit `Close`, and readonly files drop any errors on destruction.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+
+  // Close the directory base object prior to removing it.
+  static_cast<Dir&>(*this) = Dir();
+  return Cwd().RmdirRecursively(abs_path_);
```
This is an unexpected spelling for me. Why isn't it just `RmdirRecursively({})` or similar, using `*this` as the dir. Does it need to be closed first? Maybe `Rmdir()` and `RmdirRecursively()` could handle the case where you want to delete itself?

chandlerc replied:
You can't remove via a file descriptor -- the removal is about removing the name in the parent directory. So you have to have some parent directory and entry name, or you have to have a further path and remove the last component.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+    CARBON_CHECK(result == 0, "{0}",
+                 FdError(errno, "Dir::Reader::Destroy on '{0}'", fd));
+  }
+  dirp_ = nullptr;
```
this can be inside the if block

chandlerc replied:
done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  if (dirp_) {
+    int fd = dirfd(dirp_);
+    int result = closedir(dirp_);
+    // Closing a directory shouldn't produce interesting errors, so check fail
```
Same Q about why the filesystem couldn't produce an error here if e.g. there were writes in the dir queued.

chandlerc replied:
(keeping this in the thread above)

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+  // traversal back down.
+  std::optional<Dir> dir_tmp;
+  // Paths typically consist of relatively few components
+  // and so we can use a bit of stack and avoid allocated and moving the paths
```
```suggestion
  // and so we can use a bit of stack and avoid allocating and moving the paths
```
I think?

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+  std::optional<Dir> dir_tmp;
+  // Paths typically consist of relatively few components
+  // and so we can use a bit of stack and avoid allocated and moving the paths
+  // in common cases. We use `8` as an arbitrary but likely good for all of the
```
Looks like you went with 32, did that come from benchmarking?

chandlerc replied:
No, I meant to reduce it to 8. Super deep paths seem really weird.

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+  }
+  CARBON_CHECK(!missing_components.empty());
+
+  // If we haven't yet opened an intermediate directory, start by creating
```
```suggestion
  // If we haven't yet opened an intermediate directory, start by creating one
```

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+  // descriptors have to be allocated for any intervening directories, etc. We
+  // keep the path components that are missing as we pop them off for easy
+  // traversal back down.
+  std::optional<Dir> dir_tmp;
```
```suggestion
  std::optional<Dir> work_dir;
```

("tmp" has a specific meaning that doesn't apply here)

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+      // This is a directory, so try to speculatively remove it. This will fail
+      // for non-empty directories, but avoids opening, reading, and closing the
+      // directory when already empty.
+      auto rmdir_result = read_result->Rmdir(name.str());
```
There's a bunch of round trips from the entry's name which could be a `path` into a `StringRef`, then `std::string` and then `path`. Would this be more efficient if we had `Dir::Entry::name()` return a `path`, or had another version of it that did?

chandlerc replied:
Because we accept by-copied-value here in order to propagate into error objects, we don't have anything to save by re-using the string.

The result of `name` can go into either a `StringRef` or a `std::string` or a `std::filesystem::path` -- it's a C-string return because that's what Unix-like OS-es provide. =/ The `StringRef` is just caching the `strlen` result, and then we're allocating into a temporary for the by-value parameter each time.

chandlerc replied:
Note, this got cleaned up when moving back to const-ref.

danakj replied:
Yes, looks better now, thanks

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+
+  CARBON_RETURN_IF_ERROR(push_dir(*this, path));
+
+  while (true) {
```
Just noting that some of this PR uses `for (;;)` and some uses `while (true)`. I prefer the latter, but maybe we should pick one either way?

chandlerc replied:
Hmm, I prefer `while (true)`, but I think our style prefers `for (;;)`, i'll try to consolidate.

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+    if (current.dir_entries_start != static_cast<ssize_t>(dir_entries.size())) {
+      CARBON_CHECK(current.dir_entries_start <
+                   static_cast<ssize_t>(dir_entries.size()));
+      CARBON_RETURN_IF_ERROR(push_dir(current.dir, dir_entries.back()));
```
IIUC This approach ends up holding an open fd to every non-empty directory along a path in the subtree. Are we worried about hitting fd limits in weird cases with this approach, vs recursing inside each non-empty subdir without holding open all its siblings?

chandlerc replied:
The only way I see to avoid holding open a number of file descriptors equal to the max depth is to re-resolve paths as we go... If I'm missing an approach that is more conservative in open file descriptors, let me know?

The benchmark does stress test this somewhat. I was able to make things have problems, but only with pretty excessive sizes.

danakj replied:
I think the different option here is holding open one FD per depth level, instead of all siblings per depth level:

```
a
|````````````|````````````|````````````|
b            c            d            e
|```|```|    |```|```|    |```|```|    |```|```|
b1  b2  b3   c1  c2  c3   d1  d2  d3   e1  e2  e3
|
z
```

If I am understanding correctly, when we get to `z` right now, we'd have an open FD for all of `a, b, c d, e, b1, b2, b3`, but it would be possible to restructure things to only have open `a, b, b1` instead.

This came to top of mind for me because Chrome experienced out-of-fd errors before which is not a good time, and that sort of issue comes from large peak pressure (like this) suddenly introducing stress on the resource. It's okay to tell me we're not worried about it.

chandlerc replied:
Hmm, I see.

There is a weird tradeoff here:
- Hold FDs open for more siblings above your level, or
- Hold the `opendir` struct for the deepest depth

I had been worrying about the second, but I see how this puts pressure on FDs. My guess is the FDs are much more scarce, so happy to restructure it this way. Thanks for illustrating!

chandlerc replied:
After implementing, I ran benchmarks just for fun -- no measurable change in wall time or cycle time. But 15% more instructions executed with this approach. Weird, and only bad theories why.

But this is better than I expected and seems like a solid win to minimize open FDs.

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+  // until it fits. We cap this at 10 MiB to prevent egregious file system
+  // contents (or some bug somewhere) from exhausting memory.
+  constexpr ssize_t MaxSize = 10 << 20;
+  constexpr ssize_t MinSize =
+#ifdef PATH_MAX
+      PATH_MAX
+#else
+      1024
+#endif
+      ;
```
Maybe move this earlier (and name `FallbackBufferSize` or something) and use this constant for the `buffer_size` as well, instead of duplicating the constants?

chandlerc replied:
Good idea, and tried to improve the names a bit.

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+    if (large_buffer.size() >= MaxSize) {
+      return PathError(errno, "Readlink on '{0}' relative to '{1}'",
+                       std::move(path), dfd_);
+    }
```
We won't hit this error case until we try resize/grow again and we're already past MaxSize, so this could allow the buffer to be up to more like 20MiB, unlike what the comment above says. Should we be doing this comparison on the result of `NextPowerOf2` before calling `resize`?

Also, this error is using `errno` but does that make sense here? We don't seem to explain that we went over our own internal size limit here?

chandlerc replied:
No, all bugs.

Fixed to pre-compute the next buffer size, to compare that for `>` rather than `>=`, and use `ENOMEM` for the error.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+  OpenExisting = 0,
+  OpenAlways = O_CREAT,
+  CreateAlways = O_CREAT | O_TRUNC,
+  CreateNew = O_CREAT | O_EXCL,
```
IIUC this fails if the file/dir already exists? Maybe we can doc each one of these to say what it does?

chandlerc replied:
Yeah, these have subtle enough semantics, good to doc. Done.

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1374 @@
+using CreationFlags::CreateAlways;
+using CreationFlags::CreateNew;
+using CreationFlags::OpenAlways;
+using CreationFlags::OpenExisting;
```
Does `using enum CreationFlags;` work more simply to do this?

chandlerc replied:
Oooo, I had completely forgotten that went into C++20, nice!

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+
+  // If we actually created the directory, we also need to verify that the
+  // opened file descriptor continues to have the same permissions and the
+  // correct owner and group as we couldn't do the creation atomically with the
```
> and group

We check the UID but not the GID, and the comment below seems to be intentional about this

chandlerc replied:
Doh, removed inaccuracy.

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+    }
+    // Also check that the UID is the current effective UID. We don't currently
+    // verify the GID because it could come from the parent directory, so
+    // callers that need to should instead validate this themselves.
```
Could we ensure these security guarantees around permissions (or the lack of for GID) are documented on the API as well?

chandlerc replied:
Yeah... But I also want to think a bit more about this. I don't think I have all of the tradeoffs here fully correct yet. Leaving myself a FIXME until i do.

chandlerc replied:
Ok, I've gone back through this and firmed up my thinking a bit. I also write down more comments walking through the scenarios, how setgid, etc impact it. Take a look and lemme know if I'm missing anything.

A consequence is that I'm simplifying how we achieve this to focus on `CreateNew` exclusively. I've also exposed `O_NOFOLLOW` for more advanced usages, and am using that to fix a bug in the `MakeTmpDir` below. A knock-on effect of this is that `CreateDirectories` is now not more obviously making no security guarantees. This will probably make it faster and I've simplified the docs for it as well.

While I've updated the comments quite a bit here, I've also added more API/contract worded comments to the API itself. PTAL.

danakj replied:
Thanks, it looks good

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+
+auto MakeTmpDir() -> ErrorOr<RemovingDir, Error> {
+  std::filesystem::path tmpdir_path = "/tmp";
+  for (const char* tmpdir_env_name : {"TEST_TMPDIR", "TMPDIR"}) {
```
Looks like TEST_TMPDIR is a bazel thing? Let's leave a comment saying so

chandlerc replied:
Done.

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+    PrintErrorNumber(os, errno);
+    return Error(os.TakeStr());
+  }
+  CARBON_CHECK(result == tmpdir_path_buffer.data());
```
```suggestion
  CARBON_CHECK(result == tmpdir_path_buffer.data(), "`mkdtemp` used a modified path");
```

IIUC the intent here?

chandlerc replied:
I think slightly different, but will add this and tweak.

danakj commented on common/filesystem.cpp:
```diff
@@ -0,0 +1,515 @@
+  }
+
+  // Last but not least, the directory must also be empty (other than `.` and
+  // `..`). Any files here would represent a security issue from injected
```
Such symlinks could always appear later, no? I am not sure about framing this as a security issue. If it is, how does this prevent them from being added later?

chandlerc replied:
Yeah, as I mention above, I want to go back through the security side of this a bit.

chandlerc replied:
I've removed this part of the checks as I don't think it is adding any security, is very expensive, and just ends up misleading.

A bad actor can't put something here later because we successfully opened the directory that was created with `0700` permissions -- they can't mutate it, and every access relative to that root directory is guaranteed to not see any subsequent changes.

The safety of the *path* provided is dependent on the system having a secure `TMPDIR` or `TEST_TMPDIR` setup. Not sure there is any way to validate that, but achieving that is precisely what `mkdtemp` combined with `TMPDIR` are supposed to do.

danakj replied:
SGTM

danakj commented:
Did not get to the benchmark or test, but this is already a lot of feedback so stopping here for now

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

chandlerc committed:
review feedback

chandlerc reviewed (commented):
Thank you so much for the review!!!

Updated I *think* getting to all the comments to at least discuss things, and making many of the changes requested. PTAL!

chandlerc committed:
more review

chandlerc reviewed (commented):
Doh, forgot to go back to the `.cpp` file. Now with those responses as well.

chandlerc reviewed (commented):
Doh, forgot to go back to the `.cpp` file. Now with those responses as well.

danakj commented on common/filesystem.h:
```diff
@@ -63,55 +63,111 @@
 // to Windows.
+//
+// [1]: Note that we refer to platforms as "Unix-like" rather than POSIX as we
+// want
```
wrapping

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -63,55 +63,111 @@
+//      classification used in LLVM.
+//
+// [2]: The path parameters to most APIs are accepted by-copied-value in this
+// API
```
wrapping

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -230,13 +286,7 @@ class Internal::FileRefBase {
-  auto operator=(const FileRefBase&) -> FileRefBase& = default;
-
-  // Computers the file status.
+  // Computes the file status.
```
```suggestion
  // Reads the file status.
```

danakj commented on common/filesystem.h:
```diff
@@ -844,13 +932,17 @@ class DirRef::Reader : public DirRef {
     arg.dirp_ = nullptr;
+    // The directory file descriptor isn't owning, but clear it for clarity.
+    arg.dfd_ = -1;
   }
```
```suggestion
  Reader(Reader&& arg) noexcept
      // The directory file descriptor isn't owning, but clear it for clarity.
      : DirRef(std::exchange(arg.dfd_, -1)),
      dirp_(std::exchange(arg.dirp_, nullptr)) {}
```

In line with the other exchange suggestions, capturing that this one matches the same pattern

chandlerc replied:
Done.

danakj commented on common/filesystem.h:
```diff
@@ -1176,38 +1267,38 @@ inline auto DirRef::Readlink(std::filesystem::path path)
 
 inline auto DirRef::OpenReadOnly(std::filesystem::path path,
-                                 CreationFlags creation_flags,
+                                 CreationOptions creation_flags,
```
```suggestion
                                 CreationOptions creation_options,
```

Here and elsewhere

danakj commented on common/filesystem.h:
```diff
@@ -0,0 +1,1464 @@
+  // Accessor for the `errno` based error number. This is not a portable API,
+  // code using it will need to be ported to use a different API on Windows.
+  // TODO: Add a Windows-specific API for its low-level error information.
+  auto number() const -> int { return number_; }
```
This is called `errnum` everywhere else, maybe match that inside this class?

chandlerc replied:
Sure, no strong opinion.

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

chandlerc committed:
more review

chandlerc committed:
still more

chandlerc committed:
dirs are hard

chandlerc reviewed (commented):
Thanks again, updated to hopefully address review comments!

chandlerc committed:
another fix and adding more/better benchmarks

chandlerc committed:
welp, back to const ref

also format

chandlerc reviewed (commented):
Welp, much to my surprise, even after I added missing benchmarks for what should be very fast paths such as testing that a file doesn't exist, the move-in, move-into-error *consistently* underperforms. Usually its lost in the noise of doing filesystem operations but in some cases its visible even on wall clock time:

```
 BM_OpenMissing<Carbon>/real_time.. │ 👍  -1.411%      p=0.0411  │ 👍 -12.077%      p=0.000722 │ 👍  -7.204%      p=0.000425
                          baseline: │      6.049 µs  ±   2.271%  │    262.1       ±   2.222%   │    111         ±   0.001%
                        experiment: │      5.964 µs  ±   3.164%  │    230.4       ±   1.998%   │    103         ±   0.001%
```

I've switched to the more conventional argument style and removed the comment about it.

chandlerc committed:
reduce number of files

chandlerc commented:
FYI, posted #5905 to show what this look like when porting //toolchain/install over.

danakj commented on common/filesystem.h:
```diff
@@ -286,16 +283,16 @@ class Internal::FileRefBase {
   auto Seek(int64_t delta) -> ErrorOr<int64_t, FdError>;
-  auto SeekTo(int64_t pos) -> ErrorOr<int64_t, FdError>;
-  auto SeekFromEnd(int64_t delta) -> ErrorOr<int64_t, FdError>;
+  auto SeekFromBeginning(int64_t pos) -> ErrorOr<int64_t, FdError>;
```
nit:
```suggestion
  auto SeekFromBeginning(int64_t delta_from_beginning) -> ErrorOr<int64_t, FdError>;
```
Would more closely parallel the SeekFromEnd

danakj commented on common/filesystem.h:
```diff
@@ -610,34 +609,49 @@ class DirRef {
+  // replace the created directory with one that is controlled by the adversary.
+  //
+  // Specifically, for `CreateNew` we ensure that the last component is a
+  // created directory in its parent, and  cannot be replaced by a symlink into
```
```suggestion
  // created directory in its parent, and cannot be replaced by a symlink into
```

danakj commented on common/filesystem.h:
```diff
@@ -610,34 +609,49 @@ class DirRef {
+  // creation should have a single component from an opened existing parent
+  // directory. Also, no validation of the owning _group_ is performed. When
+  // securely creating a directory, the caller should either ensure the parent
+  // directory does not have a malicious setgid bit set, or restriction the
```
```suggestion
  // directory does not have a malicious setgid bit set, or restrict the
```

danakj commented on common/filesystem.h:
```diff
@@ -853,15 +864,20 @@ class DirRef::Entry {
-    return dent_->d_type == DT_UNKNOWN ? std::nullopt
-                                       : std::optional(dent_->d_type == DT_DIR);
-  }
+  // Note that we don't provide an enumerator here as we don't have any reliably
```
```suggestion
  // Note that we don't provide an enumerator here as we don't have any reliable
```

danakj commented on common/filesystem.h:
```diff
@@ -853,15 +864,20 @@ class DirRef::Entry {
 
-  // Test whether the entry is known to be a directory.
+  // Test if the entry has an unknown type. In this case, all other type
+  // predicates will return false and the caller will have to directly `lstat`
```
```suggestion
  // predicates will return false and the caller will have to directly `Lstat()`
```

danakj commented:
I think the only remaining thing that is an open discussion here is the DCHECK in the file dtor, the rest looks good now, thanks for all the improvements.

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

chandlerc committed:
tweak

chandlerc committed:
Switch to more file descriptor efficient rmtree implementation.

danakj commented:
Looks like there's something wrong in the RmTree changes, they are hitting a CHECK

chandlerc committed:
Fix to handle concurrent removal and remove check.

chandlerc commented:
> Looks like there's something wrong in the RmTree changes, they are hitting a CHECK

This is weird... it doesn't reproduce for me on either Linux machine I have, or in any build config....

There was a bug here -- we weren't handling `ENOENT` in the face of concurrent removal. I've not tried to write a test that exercises concurrent removal though, and the error number in the  `CHECK` failure is `ENOTEMPTY`. And `ENOTEMPTY` is ... *weird*.

But there was no reason to `CHECK` fail on it, we're about to propagate errors, so I've just done that with the `ENOENT` fix. But if this is consistently happening, I'd expect a `CHECK` failure up the call stack anyways, let's see...

chandlerc committed:
debugging-hacks

chandlerc committed:
try out rewinding

chandlerc committed:
replace rewinding with a similar data structure approach to original w/o excess fd

chandlerc committed:
force close on writable files and drop errors on readonly

chandlerc committed:
fix

chandlerc commented:
> > Looks like there's something wrong in the RmTree changes, they are hitting a CHECK
> 
> This is weird... it doesn't reproduce for me on either Linux machine I have, or in any build config....
> 
> There was a bug here -- we weren't handling `ENOENT` in the face of concurrent removal. I've not tried to write a test that exercises concurrent removal though, and the error number in the `CHECK` failure is `ENOTEMPTY`. And `ENOTEMPTY` is ... _weird_.
> 
> But there was no reason to `CHECK` fail on it, we're about to propagate errors, so I've just done that with the `ENOENT` fix. But if this is consistently happening, I'd expect a `CHECK` failure up the call stack anyways, let's see...

After some diagnostics, it appears that the GitHub action runner environment doesn't provide the expected guarantees when unlinking/rmdir-ing entries while reading. My understanding of the POSIX [spec](https://pubs.opengroup.org/onlinepubs/9699919799/functions/readdir.html) is that while it is unspecified whether a removed file is returned by subsequent `readdir` calls, in the action runners files that are *not* removed fail to be returned by subsequent `readdir` calls. This doesn't seem right to me, and doesn't seem to happen on macOS, or either of two Linux machines I have, but it's the only explanation.

As a consequence I've added some comments and switched back to an implementation that does a single pass over the directory and caches the entries found. Benchmarking this appears to be slightly faster than rewinding and re-reading the directory until it is confirmed empty. I *think* I have managed to preserve the minimal file descriptor usage, and it appears to be about 5% faster, so that's nice.

I've also implemented the requirement for explicit `Close` on writable files.

danakj commented on common/filesystem.cpp:
```diff
@@ -303,93 +311,96 @@ auto DirRef::CreateDirectories(const std::filesystem::path& path,
+      if (entry.is_known_dir()) {
+        dir_entries.push_back(name.str());
+      } else {
+        unknown_entries.push_back(name.str());
```
These are "unknown or files", which took a second to reason out as I couldn't find where we delete the files. Maybe a comment here or somewhere?

chandlerc replied:
Sure, added comments both here and below to try and help.

danakj commented on common/filesystem.h:
```diff
@@ -415,6 +427,14 @@ class FileRef : public Internal::FileRefBase {
+  // Note: this is a private API that should not be made public, and should only
+  // be used by the implementation of subclass destructors. It should also only
+  // be called for subclasses with *ownership* of the file reference, and is
+  // provided here as a single non-template implementation.
```
This is in a template class, so this last part doesn't hold.

Maybe subclasses should just call the right thing directly then?

chandlerc replied:
Doh, this was copy-pasted from above.

But you're right, there's no advantage to not putting this in the `File` template below. Moved it there and simplified the comment to just clarify this handles the dispatching and the core destroy logic comes from the non-template base.

danakj commented on common/filesystem.h:
```diff
@@ -1411,6 +1449,27 @@ constexpr auto Dir::Destroy() -> void {
+    //        here, there is no potentially dropped data to report.
+    //
+    // If we ever discover a platform that fails here, we should adjust this
+    // code to not fail in the face of that, likely be dropping the error. If we
```
```suggestion
    // code to not fail in the face of that, likely by dropping the error. If we
```

danakj commented on common/filesystem.h:
```diff
@@ -1411,6 +1449,27 @@ constexpr auto Dir::Destroy() -> void {
+    // code to not fail in the face of that, likely be dropping the error. If we
+    // end up supporting a platform that actually requires well-specified
+    // retries, this code should handle that. Until then, we require these to
+    // succeed so we will learn about any issues.
```
```suggestion
    // succeed so we will learn about any issues during porting to new platforms.
```

danakj reviewed (approved):
LGTM

chandlerc committed:
Apply suggestions from code review

Co-authored-by: Dana Jansens <danakj@orodu.net>

chandlerc committed:
more review

chandlerc reviewed (commented):
Thanks for the review and LGTM! I think I've addressed all the comments, but if anything needs another look, let me know and I'll send a follow-up. For now, merging!
